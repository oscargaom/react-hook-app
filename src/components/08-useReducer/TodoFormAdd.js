import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoFormAdd = ({handleAddNewTodo}) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ""
    });
    
    // console.log(description);

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length < 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddNewTodo(newTodo);

        reset();
    };

    return (
        <>
            <h4> Agregar TODO </h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Agregar nueva tarea ..."
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
