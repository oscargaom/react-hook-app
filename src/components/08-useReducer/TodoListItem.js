import React from 'react'

export const TodoListItem = ({todo, index, handleToggle, handleRemoveOnClick}) => {

    const {id, done, desc} = todo;

    return (
        <li
            key={id}
            className="list-group-item"
        >
            <p 
                className={`${done && "complete"}`}
                onClick={() => handleToggle(id)}
            >
                {index + 1}. {desc}
            </p>

            <button
                name={id}
                className="btn-danger"
                onClick={() => { handleRemoveOnClick(id) }}
            >
                Eliminar
            </button>
        </li>
    )
}
