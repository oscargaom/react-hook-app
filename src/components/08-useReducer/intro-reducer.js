
// Se crea un arreglo con un objeto.
const initialState = [{
    id: 1,
    todo: "Solfeo",
    done: false
}];

const reducer = (state = initialState, action) => {

    if (action?.type === "agregar") {
        return [...state, action.payload];
    }

    return state;
};

const todos = reducer();

const newTodo = {
    id: 2,
    todo: "Practiar por 15 minutos",
    done: false
};

const agregarTodoAction = {
    type: "agregar",
    payload: newTodo
};

const newTodos = reducer(todos, agregarTodoAction);

console.log(todos);
console.log(newTodos);
