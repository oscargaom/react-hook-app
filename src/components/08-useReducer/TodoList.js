import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, handleToggle, handleRemoveOnlick}) => {
    return (
        <ul className="list-group list-group-flush">
        {
            todos.map((todo, idx) =>
                <TodoListItem 
                    key={todo.id}
                    todo={todo} 
                    index={idx} 
                    handleToggle={handleToggle} 
                    handleRemoveOnlick={handleRemoveOnlick}
                />
            )
        }
    </ul>
    )
}
