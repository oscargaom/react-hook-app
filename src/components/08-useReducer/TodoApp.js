import React, { useReducer, useEffect } from 'react'
import "../../styles/styles.css"
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoFormAdd } from './TodoFormAdd';

// const initialState = [{
//     id: new Date().getTime(),
//     desc: "Solfeo",
//     done: false
// }];

const init = () => {
    // return [{
    //     id: new Date().getTime(),
    //     desc: "Solfeo",
    //     done: false
    // }]
    return JSON.parse(localStorage.getItem("todos")) || [];
};


export const TodoApp = () => {

    // const [todos, dispatch] = useReducer(todoReducer, initialState);
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
        // console.log('entrando a useEffect');
        // console.log(todos);
    }, [todos]);

    const handleRemoveOnClick = (todoId) => {

        // const todosJson = JSON.parse(localStorage.getItem("todos"))
        // console.log(todosJson);
        // const { target: { name } } = e;
        // console.log(name);

        const action = {
            type: "remove",
            payload: todoId
        };

        dispatch(action);
    };

    const handleToggle = (todoId) => {
        console.log(todoId);
        dispatch({
            type: "toggle",
            payload: todoId
        });
    };

    const handleAddNewTodo = (todo) => {
        
        dispatch({
            type: "add",
            payload: todo
        });
    };

    return (
        <div>
            <h1>Todo App ({todos.length}) </h1>
            <hr />

            {/* <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, idx) =>
                                (<li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p 
                                        className={`${todo.done && "complete"}`}
                                        onClick={() => handleToggle(todo.id)}
                                    >
                                        {idx + 1}. {todo.desc}
                                    </p>

                                    <button
                                        name={todo.id}
                                        className="btn-danger"
                                        onClick={() => { handleRemoveOnClick(todo.id) }}
                                    >
                                        Eliminar
                                    </button>
                                </li>)
                            )
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4> Agregar TODO </h4>
                    <hr />

                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="description"
                            className="form-control"
                            placeholder="Agregar nueva tarea ..."
                            autoComplete="off"
                            value={description}
                            onChange={handleInputChange}
                        />

                        <button
                            type="submit"
                            className="btn btn-outline-primary mt-1 btn-block"
                        >
                            Agregar
                        </button>
                    </form>
                </div>
            </div> */}
            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleToggle={handleToggle}
                        handleRemoveOnClick={handleRemoveOnClick}
                    />
                </div>
                <div className="col-5">
                    <TodoFormAdd 
                        handleAddNewTodo={handleAddNewTodo}
                    />
                </div>
            </div>
        </div>
    )
}
