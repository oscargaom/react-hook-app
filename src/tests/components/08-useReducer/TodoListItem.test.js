import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { newDemoTodoItem } from '../../fixtures/demoTodos';

describe('Pruebas del componente <TodoListItem />', () => {

    const handleRemoveOnClick = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(<TodoListItem
        todo={newDemoTodoItem}
        index={newDemoTodoItem.id}
        handleToggle={handleToggle}
        handleRemoveOnClick={handleRemoveOnClick}
    />);

    test('Debe mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de llamar la funci칩n handleRemoveOnClick', () => {

        wrapper.find('button').simulate('click');
        
        expect(handleRemoveOnClick).toHaveBeenCalled();
        expect(handleRemoveOnClick).toHaveBeenCalledTimes(1);
        // V치lida que se haya llamado con el valor del par치metro newDemoTodoItem.id
        expect(handleRemoveOnClick).toHaveBeenCalledWith(newDemoTodoItem.id);
        // expect(handleToggle).not.toHaveBeenCalled();
        // expect(handleToggle).toHaveBeenCalledTimes(0);
    });

    test('Debe de llamar la funci칩n handleToggle', () => {
        
        wrapper.find('p').simulate('click');
        
        expect(handleToggle).toHaveBeenCalled();
        expect(handleToggle).toBeCalledTimes(1);
        expect(handleToggle).toHaveBeenCalledWith(newDemoTodoItem.id);
    });

    test('Debe mostrar el texto correctamente', () => {
        const parrafo = wrapper.find('p').text().trim();

        expect(parrafo).toBe(`${newDemoTodoItem.id + 1}. ${newDemoTodoItem.desc}`);
    });

    test('Debe tener el parrafo la clase complete si done es true ', () => {

        //const handleRemoveOnClick2 = jest.fn();
        //const handleToggle2 = jest.fn();

        jest.clearAllMocks();

        newDemoTodoItem.done = true;

        const wrapper2 = shallow(<TodoListItem
                                    todo={newDemoTodoItem}
                                    index={newDemoTodoItem.id}
                                    handleToggle={handleToggle}
                                    handleRemoveOnClick={handleRemoveOnClick}
                                />);
        // console.log(wrapper2.find('p').hasClass('complete'));
        expect(wrapper2.find('p').hasClass('complete')).toBe(true);
    });
})
