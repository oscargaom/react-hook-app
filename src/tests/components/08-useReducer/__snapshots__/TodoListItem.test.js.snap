// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <TodoListItem /> Debe mostrarse correctamente 1`] = `
<li
  className="list-group-item"
  key="9"
>
  <p
    className="false"
    onClick={[Function]}
  >
    10
    . 
    Dormir
  </p>
  <button
    className="btn-danger"
    name={9}
    onClick={[Function]}
  >
    Eliminar
  </button>
</li>
`;
